In [1]:
import pandas as pd
pd.options.display.float_format = '{:,.0f}'.format  # Format numbers

In [2]:
df = pd.read_csv("Micah's CS Projects/GenAI Financial Chatbot (BCG)/10-X Form Analysis.csv")

In [3]:
# Sort the rows by company and year in ascending order
df = df.sort_values(["Company", "Year"])
df = df.reset_index(drop=True)

In [4]:
# Show column names
print(df.columns.tolist())

['Company', 'Year', 'Total Revenue', 'Net Income', 'Total Assets', 'Total Liabilities', 'Cash Flow from Operating Activities']


In [5]:
# Represent all numerical values as floats
df['Total Revenue'] = (
    df['Total Revenue']
    .astype(str)                      # Force all values to string
    .str.replace(',', '', regex=False)  # Remove commas
    .astype(float)                   # Convert to float
)
df['Net Income'] = (
    df['Net Income']
    .astype(str)                      # Force all values to string
    .str.replace(',', '', regex=False)  # Remove commas
    .astype(float)                   # Convert to float
)
df['Total Assets'] = (
    df['Total Assets']
    .astype(str)                      # Force all values to string
    .str.replace(',', '', regex=False)  # Remove commas
    .astype(float)                   # Convert to float
)
df['Total Liabilities'] = (
    df['Total Liabilities']
    .astype(str)                      # Force all values to string
    .str.replace(',', '', regex=False)  # Remove commas
    .astype(float)                   # Convert to float
)
df['Cash Flow from Operating Activities'] = (
    df['Cash Flow from Operating Activities']
    .astype(str)                      # Force all values to string
    .str.replace(',', '', regex=False)  # Remove commas
    .astype(float)                   # Convert to float
)

Here are the columns that will be analyzed within the data from the 10-K forms.

In [6]:
# Display the original values before calculations
df.head(9)

Company  Year   Total Revenue     Net Income    Total Assets  \
0             Apple Inc.  2022 394,328,000,000 99,803,000,000 352,755,000,000   
1             Apple Inc.  2023 383,285,000,000 96,995,000,000 352,583,000,000   
2             Apple Inc.  2024 391,035,000,000 93,736,000,000 364,980,000,000   
3  Microsoft Corporation  2022 198,270,000,000 72,738,000,000 364,840,000,000   
4  Microsoft Corporation  2023 211,915,000,000 72,361,000,000 411,976,000,000   
5  Microsoft Corporation  2024 245,122,000,000 88,136,000,000 512,163,000,000   
6            Tesla, Inc.  2022  81,462,000,000 12,587,000,000  82,338,000,000   
7            Tesla, Inc.  2023  96,773,000,000 14,997,000,000 106,618,000,000   
8            Tesla, Inc.  2024  97,690,000,000  7,153,000,000 122,070,000,000   

   Total Liabilities  Cash Flow from Operating Activities  
0    302,083,000,000                      122,151,000,000  
1    290,437,000,000                      110,543,000,000  
2    308,030,000,000                      118,254,000,000  
3    198,298,000,000                       89,035,000,000  
4    205,753,000,000                       87,582,000,000  
5    243,686,000,000                      118,548,000,000  
6     36,440,000,000                       14,724,000,000  
7     43,009,000,000                       13,256,000,000  
8     48,390,000,000                       14,923,000,000

Here is the orginal data before any calculations. The table is organized by company and the years are sorted in ascending order. Following the year and company are the Total Revenue, Net Income, Total Assets, Total Liabilities, and the Cash Flow from Operating Activities.

In [7]:
# Create a column to represent Revenue Growth
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100

In [8]:
# Create a column to represent Net Income Growth
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [9]:
# Create a column to represent Net Margin
df['Net Margin'] = df['Net Income'] / df['Total Revenue']

In [10]:
# Create a column to represent Debt to Asset Ratio
df['Debt to Asset Ratio'] = df['Total Liabilities'] / df['Total Assets']

In [11]:
# Create a column to represent Asset Growth
df['Asset Growth (%)'] = df.groupby('Company')['Total Assets'].pct_change() * 100

In [12]:
# Create a column to represent Liability Growth
df['Liability Growth (%)'] = df.groupby('Company')['Total Liabilities'].pct_change() * 100

In [13]:
# Create a column to represent Return on Assets (ROA)
df['Return on Assets'] = df['Net Income'] / df['Total Assets']

In [14]:
# Create a column to represent Cash Flow Growth (%)
df['Cash Flow Growth (%)'] = df.groupby('Company')['Cash Flow from Operating Activities'].pct_change() * 100

In [15]:
# Organize table values
df = df[[
    'Company','Year',
    'Total Revenue','Revenue Growth (%)',
    'Net Income','Net Income Growth (%)','Net Margin',
    'Total Assets','Asset Growth (%)','Return on Assets','Debt to Asset Ratio',
    'Total Liabilities','Liability Growth (%)',
    'Cash Flow from Operating Activities', 'Cash Flow Growth (%)'
]]

In [16]:
# Format percent columns and ratios
pd.options.display.float_format = '{:,.3f}'.format  # Format numbers
df['Total Revenue'] = df['Total Revenue'].apply(lambda x: f"{x:,.0f}")
df['Net Income'] = df['Net Income'].apply(lambda x: f"{x:,.0f}")
df['Total Assets'] = df['Total Assets'].apply(lambda x: f"{x:,.0f}")
df['Total Liabilities'] = df['Total Liabilities'].apply(lambda x: f"{x:,.0f}")
df['Cash Flow from Operating Activities'] = df['Cash Flow from Operating Activities'].apply(lambda x: f"{x:,.0f}")

In [17]:
# Display table
df.head(9)

Company  Year    Total Revenue  Revenue Growth (%)  \
0             Apple Inc.  2022  394,328,000,000                 NaN   
1             Apple Inc.  2023  383,285,000,000              -2.800   
2             Apple Inc.  2024  391,035,000,000               2.022   
3  Microsoft Corporation  2022  198,270,000,000                 NaN   
4  Microsoft Corporation  2023  211,915,000,000               6.882   
5  Microsoft Corporation  2024  245,122,000,000              15.670   
6            Tesla, Inc.  2022   81,462,000,000                 NaN   
7            Tesla, Inc.  2023   96,773,000,000              18.795   
8            Tesla, Inc.  2024   97,690,000,000               0.948   

       Net Income  Net Income Growth (%)  Net Margin     Total Assets  \
0  99,803,000,000                    NaN       0.253  352,755,000,000   
1  96,995,000,000                 -2.814       0.253  352,583,000,000   
2  93,736,000,000                 -3.360       0.240  364,980,000,000   
3  72,738,000,000                    NaN       0.367  364,840,000,000   
4  72,361,000,000                 -0.518       0.341  411,976,000,000   
5  88,136,000,000                 21.800       0.360  512,163,000,000   
6  12,587,000,000                    NaN       0.155   82,338,000,000   
7  14,997,000,000                 19.147       0.155  106,618,000,000   
8   7,153,000,000                -52.304       0.073  122,070,000,000   

   Asset Growth (%)  Return on Assets  Debt to Asset Ratio Total Liabilities  \
0               NaN             0.283                0.856   302,083,000,000   
1            -0.049             0.275                0.824   290,437,000,000   
2             3.516             0.257                0.844   308,030,000,000   
3               NaN             0.199                0.544   198,298,000,000   
4            12.920             0.176                0.499   205,753,000,000   
5            24.319             0.172                0.476   243,686,000,000   
6               NaN             0.153                0.443    36,440,000,000   
7            29.488             0.141                0.403    43,009,000,000   
8            14.493             0.059                0.396    48,390,000,000   

   Liability Growth (%) Cash Flow from Operating Activities  \
0                   NaN                     122,151,000,000   
1                -3.855                     110,543,000,000   
2                 6.057                     118,254,000,000   
3                   NaN                      89,035,000,000   
4                 3.759                      87,582,000,000   
5                18.436                     118,548,000,000   
6                   NaN                      14,724,000,000   
7                18.027                      13,256,000,000   
8                12.511                      14,923,000,000   

   Cash Flow Growth (%)  
0                   NaN  
1                -9.503  
2                 6.976  
3                   NaN  
4                -1.632  
5                35.357  
6                   NaN  
7                -9.970  
8                12.575

As you can see here, I have created more columns to represent the Revenue Growth, Net Income Growth, Net Margin, Asset Growth, Liability Growth, Debt to Asset Ratio, and Return on Assets.

For Apple Inc., during 2022-2023, the total revenue decreased by 2.8%. Apple's net income also decreased by 2.8%. As for Apple's assets, they decreased by 0.05%, while Apple's liability value decreased by 3.9%. Apple's cash flow from operating activities decreased by 9.5%. The net margin for 2022 and 2023 was 0.253 and the return on assets in 2022 was 0.283 and in 2023 it was 0.275. During 2022 the debt to asset ratio was 0.856, while in 2023 it was slightly lower at 0.824. During the following years, 2023-2024, Apple's total revenue increased by 2% and its net income decreased by 3.4%. Apple's asset value increased by 3.5% and Apple's liability value grew by 6.1%. In 2024, cash flow from Operating Activities grew by about 7%. The net margin for 2024 was 0.24 and the return on assets was 0.257. Apple's debt to asset ratio in 2024 was 0.844.

Microsoft Corporation earned less revenue than Apple Inc. but showed revenue growth. During 2022-2023, Microsoft Corporation's revenue increased by 6.9%. In the same year, Microsoft Corporation's net income decreased by 0.5%. Microsoft's assets grew by 12.9% and liabilities grew by 3.8%. Cash flow from operating activities decreased by 1.6%. The net margin for Microsoft Corporation was 0.367 in 2022 and 0.341 in 2023. Microsoft's ROA for 2022 was 0.199 and 0.176 in 2023. In 2022, Microsoft's debt to asset ratio was 0.544 and in 2023 it was 0.499. During the 2023-2024 period, Microsoft experienced even more growth in revenue (15.7%), net income (21.8%), assets (24.3%), and liabilties (18.4%). Cash flow from operating activities grew significatly by 35.4%. The net margin for 2024 was 0.34 and the ROA for the same year was 0.172. The debt to asset ratio in 2024 was 0.476.

Tesla earned less than both Apple and Microsoft and they lost a significant amount of income in 2024. During the 2022-2023 period, Tesla's revenue grew by 18.8% and Tesla's net income increased by 19.1%. Tesla's assets grew by 29.5% and their liabilities grew by 18%. Cash flow from operating activities decreased by 10%. Tesla's net margin in 2022 and 2023 was 0.155. And Tesla's return on assets was 0.153 in 2022 and 0.141 in 2023. Tesla's debt to asset ratio was 0.443 in 2022 and 0.403 in 2023. During the following period, 2023-2024, Tesla's total revenue grew by 0.948 but Tesla's net income decreased significantly by 52.3%. In 2024, Tesla's assets grew by 14.5% and Tesla's liabilities increased by 12.5%. Tesla's cash flow from operating activities grew by 12.6%. The net margin for 2024 was 0.073 and its return on assets was 0.059. The debt to asset ratio in 2024 was 0.396.